### ISSUE TITLE
Can not install Ripple-qewd with GT.M as its database

### ISSUE THREAD

**Issue by maxcda101** [RippleOSI/Ripple-Qewd#46]
Hi @robtweed 
I have problems while installing Ripple-qewd with GT.M
This is my logs when i running ripple-demo.js or ripple-secure.js
https://prnt.sc/frt0sn
Help me point out the problem.
Thanks

Quan - From ThinkLabs.vn

**Comment by robtweed** [comment]
I just tested the qewd-ripple + GT.M installer on an Amazon Web Services Ubuntu 16.04 machine - it installed and ran OK.

I'd need to see your installation log, but it looks like the Node.js interface for GT.M (NodeM) hasn't been installed correctly

Can you give details of the configuration you're trying to use and how you're installing?  Are you using Ubuntu 16.04 and are you using the installation script we provide?

**Comment by maxcda101** [comment]
I'm using Ubuntu 16.04 on docker image https://hub.docker.com/_/ubuntu-upstart and user root.
I installed success qewd-ripple + Redis, but GT.M not yet.

**Comment by robtweed** [comment]
You'll see in the installation ReadMe document that you should not install qewd-ripple as root.  I would guess that it's a GT.M installation permissions issue.  I know you're using a Docker image, but if you can set up a non-root user with sudo privileges, I think you'll find it a lot easier to get it up and working

**Comment by maxcda101** [comment]
I created a new image docker and  user ripple then reinstall but the same error old, this is installation log 
[log.txt](https://github.com/RippleOSI/Ripple-Qewd/files/1124504/log.txt)

**Comment by robtweed** [comment]
Just to check - it has to be a 64-bit version of Ubuntu 16.04.  It appears to be the GT.M installation that is failing - I'll see what I can find out

**Comment by robtweed** [comment]
It appears that the GT.M installation is failing because it can't create the /usr/lib/fis-gtm directory.  Here's the feedback I had from a GT.M guru.  The key line in the installation log is this:

ls: cannot access /usr/lib/fis-gtm/V6.3-001A_x86_64: No such file or directory

"It appears that that path is not available in the docker container (I would hazard a guess that it's not allowing an fis-gtm subdirectory in /usr/lib)."

I think you'll need to look at the way the Ubuntu Docker container is set up in order to figure out what to do to make it install GT.M.  According to others who work with GT.M all the time, they tell me they have successfully installed GT.M on Docker containers and there's nothing usually problematic.  So I guess it must be something specific to the container you're using

Hope this helps!
Rob

**Comment by tony-shannon** [comment]
@maxcda101 Can I get an update from you on this issue please.. is this resolved? 
Can we close off this issue?
T

**Comment by thanh121094** [comment]
Hi @robtweed, @tony-shannon  . I'm using Desktop Ubuntu 16.04 - 64 bit machine. I have some problems while installing Ripple-qewd with GT.M. This is my logs when I run pm2 log ripple-demo.
````
$ pm2 log ripple-demo
[TAILING] Tailing last 15 lines for [ripple-demo] process (change the value with --lines option)
/home/thanhnv/.pm2/logs/ripple-demo-error-0.log last 15 lines:
0|ripple-d | path.js:1149
0|ripple-d |           cwd = process.cwd();
0|ripple-d |                         ^
0|ripple-d | 
0|ripple-d | Error: ENOENT: no such file or directory, uv_cwd
0|ripple-d |     at Error (native)
0|ripple-d |     at Object.resolve (path.js:1149:25)
0|ripple-d |     at startup (bootstrap_node.js:127:32)
0|ripple-d |     at bootstrap_node.js:496:3
0|ripple-d | Inappropriate ioctl for device
0|ripple-d | %GTM-E-GTMDISTUNDEF, Environment variable $gtm_dist is not defined
0|ripple-d | Inappropriate ioctl for device
0|ripple-d | %GTM-E-GTMDISTUNDEF, Environment variable $gtm_dist is not defined
0|ripple-d | Inappropriate ioctl for device
0|ripple-d | %GTM-E-GTMDISTUNDEF, Environment variable $gtm_dist is not defined

/home/thanhnv/.pm2/logs/ripple-demo-out-0.log last 15 lines:
0|ripple-d | process.argv[2] = qewd.worker
0|ripple-d | workerModule: qewd; worker
0|ripple-d | *** master received exit event from worker process 7719
0|ripple-d | *** CTRL & C detected: shutting down gracefully...
0|ripple-d | No worker processes are running
0|ripple-d | Master process will now shut down
0|ripple-d | webServerRootPath = /home/thanhnv/qewd/www
0|ripple-d | route /api will be handled by qx.router
0|ripple-d | Worker Bootstrap Module file written to node_modules/ewd-qoper8-worker.js
0|ripple-d | ========================================================
0|ripple-d | ewd-qoper8 is up and running.  Max worker pool size: 2
0|ripple-d | ========================================================
0|ripple-d | ========================================================
0|ripple-d | QEWD.js is listening on port 3000
0|ripple-d | ========================================================

0|ripple-d | no available workers
0|ripple-d | sent qoper8-start message to 14823
0|ripple-d | process.argv[2] = qewd.worker
0|ripple-d | workerModule: qewd; worker
0|ripple-d | Inappropriate ioctl for device
0|ripple-d | %GTM-E-GTMDISTUNDEF, Environment variable $gtm_dist is not defined
0|ripple-d | *** master received exit event from worker process 14823
0|ripple-d | no available workers
0|ripple-d | sent qoper8-start message to 14853
0|ripple-d | process.argv[2] = qewd.worker
0|ripple-d | workerModule: qewd; worker
0|ripple-d | Inappropriate ioctl for device
0|ripple-d | %GTM-E-GTMDISTUNDEF, Environment variable $gtm_dist is not defined
0|ripple-d | *** master received exit event from worker process 14853
0|ripple-d | no available workers
0|ripple-d | sent qoper8-start message to 14901
0|ripple-d | process.argv[2] = qewd.worker
0|ripple-d | workerModule: qewd; worker
0|ripple-d | Inappropriate ioctl for device
0|ripple-d | %GTM-E-GTMDISTUNDEF, Environment variable $gtm_dist is not defined
0|ripple-d | *** master received exit event from worker process 14901
````

> It appears that the GT.M installation is failing because it can't create the /usr/lib/fis-gtm directory. 

However, when I checked my project, it has already created that file.


![screenshot from 2017-10-17 17-39-27](https://user-images.githubusercontent.com/24429197/31660634-3bb20bee-b362-11e7-8c6e-22d7be359498.png)

Do you have any ideas about above issues?
Thanks

Thanh - From ThinkLabs - VN

**Comment by robtweed** [comment]
The problem arises because you're using Docker, and there's no .profile file to establish the correct GT.M run-time environment automatically in the QEWD worker processes when the start.  I've seen a similar issue when running QEWD in a crouton environment on a ChromeBook.  The fix there was pretty simple, and I think it should also work in a Docker container.

Simply find your QEWD startup file (eg ~/qewd/ripple-demo.js)

Edit the config object to include the following:
    database: {
      type: 'gtm',
      params: true
    }

That params: true property will force it, when starting a new worker process, to run a special GT.M-specific module that sets up what should usually be the correct process environment variables for GT.M.

Let me know if that does the trick

Rob

**Comment by thanh121094** [comment]
Hi @robtweed , I've read your comment and I installed GT.M succesful. 

> You'll see in the installation ReadMe document that you should not install qewd-ripple as root. I would guess that it's a GT.M installation permissions issue.

Firstly, I created a new user named "ripple" then I switch to this user:
`su - ripple`
After that, I gave the user (ripple in this example) sudo privileges:
` usermod -aG sudo ripple`
Finally, I installed qewd-ripple with GT.M as its database
`  cd ~
  wget https://raw.githubusercontent.com/RippleOSI/Qewd-Ripple/master/installer/install_ripple_gtm.sh
  source install_ripple_gtm.sh`
It's ran.

Thanks for your support.

Thanh - From ThinkLabs - VN
